pam/pam.go

```diff
@@ -73,68 +83,69 @@ func pam_sm_authenticate(pamh *C.pam_handle_t, flags, argc C.int, argv **C.char)
 	p := tea.NewProgram(&appState, opts...)
 	if _, err := p.Run(); err != nil {
 		log.Errorf(context.TODO(), "Cancelled authentication: %v", err)
-		return C.PAM_ABORT
+		return pam.ErrAbort
 	}
 
-	logErrMsg := "unknown"
-	var errCode C.int = C.PAM_SYSTEM_ERR
+	var status = pam.NewTransactionError(pam.ErrSystem, errors.New("unknown exit code"))
 
-	switch exitMsg := appState.exitMsg.(type) {
+	switch exitMsg := appState.exitStatus.(type) {
 	case pamSuccess:
-		brokerIDUsedToAuthenticate = exitMsg.brokerID
-		return C.PAM_SUCCESS
+		if err := mTx.SetData(authenticationBrokerIDKey, exitMsg.brokerID); err != nil {
+			return err
+		}
+		return nil
 	case pamIgnore:
 		// localBrokerID is only set on pamIgnore if the user has chosen local broker.
-		brokerIDUsedToAuthenticate = exitMsg.localBrokerID
-		if exitMsg.String() != "" {
-			log.Debugf(context.TODO(), "Ignoring authd authentication: %s", exitMsg)
-		}
-		logErrMsg = ""
-		errCode = C.PAM_IGNORE
-	case pamAbort:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("cancelled authentication: %s", exitMsg)
-		}
-		errCode = C.PAM_ABORT
-	case pamAuthError:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("authentication: %s", exitMsg)
-		}
-		errCode = C.PAM_AUTH_ERR
-	case pamSystemError:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("system: %s", exitMsg)
+		if err := mTx.SetData(authenticationBrokerIDKey, exitMsg.localBrokerID); err != nil {
+			return err
 		}
-		errCode = C.PAM_SYSTEM_ERR
+		status = exitMsg
+	case pamReturnStatus:
+		status = exitMsg
 	}
 
-	if logErrMsg != "" {
-		fmt.Fprintf(os.Stderr, "Error: %v\n", logErrMsg)
+	if status.Status() != pam.ErrIgnore {
```

Comment:
So, `Error()` is a message that can be embedded with the error, and also `pamSuccess` or `pamIgnore` that are not errors per se could provide one. And if a message is provided, then we should send it back to the app.

Now, this could be implemented differently maybe using another method name, but since we had already a method providing an informative string, I thought was not needed to duplicate that.